index.js

```diff
@@ -407,56 +404,106 @@ class K8sExecutor extends Executor {
-                json: true
-            };
-            let res = await this.breaker.runCommand(statusOptions);
-
-            this.checkPodResponse(res, buildId);
+            const { status, nodeName, podName, waitingReason } = await this.getPodStatus({
+                podName: resp.body.metadata.name,
+                buildId,
+                retryStrategyFn: this.scheduleStatusRetryStrategy
+            });
+            const isPodInitializing = this.checkPodResponse({ status, waitingReason, buildId, podName });
             const updateConfig = {
                 apiUri: this.ecosystem.api,
                 buildId,
                 token
             };
 
-            if (res.body.spec && res.body.spec.nodeName) {
-                updateConfig.stats = {
-                    hostname: res.body.spec.nodeName,
-                    imagePullStartTime: new Date().toISOString()
-                };
+            if (nodeName) {
+                Object.assign(updateConfig, {
+                    stats: {
+                        hostname: nodeName,
+                        imagePullStartTime: new Date().toISOString()
+                    }
+                });
             } else {
-                updateConfig.statusMessage = 'Waiting for resources to be available.';
+                Object.assign(updateConfig, { statusMessage: 'Waiting for resources to be available.' });
             }
+
             await this.updateBuild(updateConfig);
-            await new Promise(r => setTimeout(r, this.podStatusQueryDelay));
-            statusOptions = {
-                uri: `${this.podsUrl}/${podName}/status`,
-                method: 'GET',
-                headers: { Authorization: `Bearer ${this.token}` },
-                strictSSL: false,
-                maxAttempts: this.maxAttempts,
-                retryDelay: this.retryDelay,
-                retryStrategy: this.pendingStatusRetryStrategy,
-                json: true
-            };
-            res = await this.breaker.runCommand(statusOptions);
-            this.checkPodResponse(res, buildId);
 
-            return null;
+            return (
+                isPodInitializing &&
+                (await this.callWithRetry(
+                    async args => {
+                        const data = await this.getPodStatus(args);
+
+                        return this.checkPodResponse(data);
+                    },
+                    { podName, buildId, retryStrategyFn: this.pendingStatusRetryStrategy }
+                ))
+            );
         } catch (err) {
             logger.error(`Failed to run pod for build id:${buildId}: ${err.message}`);
             throw err;
         }
     }
 
+    /**
+     *
+     * @param {Function} fn the fn to execute
+     * @param {Object} args  the args of the fn
+     * @param {Number} limit max limit of retries
+     * @returns {Promise} resolves to null or error
+     */
+    async callWithRetry(fn, args, limit = 0) {
```

Comment:
we already have fusebox for this retry logic. it's already used in this library https://github.com/screwdriver-cd/executor-k8s/blob/a69706a7e7e1a9dd8af8d0c26f15995ae0e1b042/index.js#L372

Reply:
yeah, that would eat up the error, can't use it here, also it's already part of request-retry, however, we still see the issue